defaults/linuxrc

```diff
@@ -656,10 +665,23 @@ then
 	start_sshd
 fi
 
+keyctl_keyadd
+
 # Initialize LUKS root device except for livecd's
 if [ "${CDROOT}" != '1' ]
 then
-	start_LUKS
+	if ( [ -n "${CRYPT_SWAP_KEY}" ] && [ -z "${CRYPT_SWAP_KEYDEV}" ] ) || \
+	   ( [ -n "${CRYPT_SWAP_HEADER}" ] && [ -z "${CRYPT_SWAP_HEADERDEV}" ] )
+	then
+		# the swap key or header might be on the root fs so start it first in this case
+		start_LUKS_root
+		luks_root_started=1
+		start_LUKS_swap
+	else
+		# we don't need to start the root at all if we are resuming from suspend
+		start_LUKS_swap
+	fi
+
```

Comment:
Some assumptions are made about configuration in this code block.  You aren't accounting for: swapfile on root, root and swap in lvm volume, etc.  These would all need LUKS_root opened before running do_resume.  Technically I suppose you could define a crypt_swap_key value the same as your root value, but that seems like misleading configuration.

Could maybe add:
`( [ "${REAL_ROOT}" = "${REAL_RESUME}" ]  || [ ${USE_LVM_NORMAL} -eq 1 ] )`

This would solve both cases (I think).

Reply:
> swapfile on root,

Looking at the relevant code in `swsusp_resume()` from `defaults/initrd.scripts` only swap partitions are supported for resume from hibernation.

You might be correct with respect to LVM though.

Reply:
I beg to differ, I'm currently running a 'LUKS -> btrfs -> swapfile' setup which resumes from hibernation just fine.  I just patched out the code-block above to enforce root LUKS being opened before resume is run, and I have 'resume=' and 'resume_offset=' set in my kernel cmdline.

Reply:
I see, you are using genkernel to write the resume device path to `/sys/power/resume` but at the same time providing the kernel directly with the swap file offset via the `resume_offset` kernel parameter. 

Will you open a PR with your proposed change then?

Reply:
Sure, I can do that.